### PR TITLE
Add cuda-11.3+clang9 build workflow (take 2)

### LINKED ISSUE #75475: PyTorch can not be compiled with CUDA acceleration using Clang as host compiler
### 🐛 Describe the bug

Running `CC=clang CXX=clang++ python setup.py develop` would fail to compile PyTorch with CUDA using clang as host compiler (as it would not be passed as ccbin option)

### Versions

N/A

### PR DESCRIPTION
**Pull request by malfet**
To be able to detect unused captures in GPU code lambdas (as gcc does not support this diagnostic)

Remove unused opts lambda capture in `ProcessGroupMPI.cpp` and `Distributions.cu`

Fix sign-compare in nvfuser benchmark and ignore signed unsigned comparison in nvfuser tests
Fixes https://github.com/pytorch/pytorch/issues/75475 by aliasing CMAKE_CUDA_HOST_COMPILER to C_COMPILER when clang is used

### CODE REVIEW

**Review by atalman — APPROVED** [review]
LGTM

File: .github/scripts/generate_ci_workflows.py [hunk 1]
```diff
@@ -667,6 +667,15 @@ def generate_workflow_file(self, workflow_template: jinja2.Template) -> None:
             labels=set([LABEL_CIFLOW_DEFAULT, LABEL_CIFLOW_LINUX, LABEL_CIFLOW_CUDA]),
         ),
     ),
+    CIWorkflow(
```
Comment (seemethere) [inline .github/scripts/generate_ci_workflows.py:670]: I don't believe we actually need to do this anymore

**Review by seemethere — APPROVED** [review]
LGTM, but we probably don't need the `generate_ci_workflows.py` change

File: .github/scripts/generate_ci_workflows.py (hunk 1 above)
Comment (malfet) [reply .github/scripts/generate_ci_workflows.py:670]: We do, otherwise it fails shellcheck lint

Comment (malfet) [reply .github/scripts/generate_ci_workflows.py]: Not needed after rebase

**Pull request closed without merging**